### PR TITLE
feat: Add Stylus Caching Strategy Concept Page

### PR DESCRIPTION
**Pull request by mahsamoosavi**

### CODE REVIEW

File: arbitrum-docs/stylus/concepts/stylus-cache-manager.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+To ensure that developers periodically pay to maintain their position in the cache, we use a global decay parameter computed by `decay = block.timestamp * _decay`. This inflates the value of bids over time, making newer bids more valuable.
+
+### Cache access and costs
+
+During activation, we compute initialization costs for both non-cached and cached initialization of the contract. These costs take into account factors such as the number of functions, types, code size, data length, and memory usage. It's important to note that ccessing an uncached contract does not automatically add it to the CacheManager's cache. Only explicit calls to the CacheManager contract will add a contract to the cache. If a contract is removed from the cache, calling the contract becomes more expensive unless it is re-added.
```
Comment (leeederek) [inline arbitrum-docs/stylus/concepts/stylus-cache-manager.md:31]:
```suggestion
During activation, we compute initialization costs for both non-cached and cached initialization of the contract. These costs take into account factors such as the number of functions, types, code size, data length, and memory usage. It's important to note that accessing an uncached contract does not automatically add it to the CacheManager's cache. Only explicit calls to the CacheManager contract will add a contract to the cache. If a contract is removed from the cache, calling the contract becomes more expensive unless it is re-added.
```

File: arbitrum-docs/stylus/concepts/stylus-cache-manager.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import PublicPreviewBannerPartial from '../partials/_stylus-public-preview-banner-partial.md';
+
+<PublicPreviewBannerPartial />
+
+## CacheManager contract
+
+The core component of our caching strategy is the [CacheManager contract](https://github.com/OffchainLabs/stylus-contracts/blob/c51ab1dc90f543caf579600162f77e053984b8cd/src/chain/CacheManager.sol). This smart contract manages the cache, interacts with precompiles, and determines which contracts should be cached. The CacheManager can hold approximately 4,000 contracts in memory.
+
+The CacheManager defines how contracts remain in the cache and how they compete with other contracts for cache space. Its primary purpose is to manage high initialization costs, ensuring efficient contract activation and usage. The contract includes methods for adding and removing cache entries, querying the status of cached contracts, and managing the lifecycle of cached data.
```
Comment (leeederek) [inline arbitrum-docs/stylus/concepts/stylus-cache-manager.md:19]:
```suggestion
The CacheManager defines how contracts remain in the cache and how they compete with other contracts for cache space. Its primary purpose is to reduce high initialization costs, ensuring efficient contract activation and usage. The contract includes methods for adding and removing cache entries, querying the status of cached contracts, and managing the lifecycle of cached data.
```

File: arbitrum-docs/stylus/concepts/stylus-cache-manager.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+### Key features
+
+The CacheManager plays a crucial role in our caching strategy by keeping a certain set of contracts in-memory rather than retrieving them from disk. This significantly reduces the activation time for frequently accessed contracts. The CacheManager contract is an on-chain contract that accepts bids for inserting contract code into the cache. It then calls a precompile that loads or unloads the contracts in the ArbOS cache, which follows the on-chain cache but operates locally in the client, and marks the contract as in or out of the cache in the ArbOS state.
```
Comment (leeederek) [inline arbitrum-docs/stylus/concepts/stylus-cache-manager.md:23]: Is it true that the cache manager reduces only activation time? Wouldn't subsequent calls (after activation) be cheaper in gas too?

File: arbitrum-docs/stylus/concepts/stylus-cache-manager.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+---
+title: 'Stylus caching strategy'
+description: 'A conceptual overview of the Stylus caching strategy and CacheManager contract, and explaining its functionality.'
+sme: mahsa-moosavi
+target_audience: 'Developers deploying smart contracts using Stylus.'
+sidebar_position: 3
+---
+
+Stylus is designed for fast computation, but it faces significant performance issues due to the slow activation process when entering a contract. This slow [activation](https://docs.arbitrum.io/stylus/stylus-gentle-introduction#activation) takes up substantial resources and time. These repeated activations lead to inefficiencies and hinder the overall performance of the system. To address this problem, we have implemented a caching strategy. The cache serves as a bypass, storing frequently accessed contracts in-memory. This approach allows for avoiding repeated activations, thereby saving resources and time while significantly improving the speed and efficiency of contract execution.
```
Comment (tsahee) [inline arbitrum-docs/stylus/concepts/stylus-cache-manager.md:9]: Activation is different. You only need to activate a program once and then it stays "active" for a year or until stylus version changes. It's unrelated to caching.
Here we're talking about something else - in some places it's called "initialization" in this document, which is a good name for it.
Most places where you describe "activation" in this doc should be changed to "initialization"

File: arbitrum-docs/stylus/concepts/stylus-cache-manager.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+The cache operates through an auction system where dapp developers submit bids to insert their contracts into the cache. If the cache is at capacity, lower bids are evicted to make space for higher bids. The cache maintains a min heap of bids for codeHashes, with bids encoded as `bid << 64 + index`, where `index` represents the position in the list of all bids. When an insertion exceeds the cache's maximum size, items are popped off the min heap and deleted until there is enough space to insert the new item. Zero bids are the first to be popped off based on their index, leading to random evictions due to the non-meaningful ordering of entries. Otherwise, the smallest bid is evicted first.
```
Comment (tsahee) [inline arbitrum-docs/stylus/concepts/stylus-cache-manager.md:25]: I don't think we need to detail zero-bids specifically. Saying that contracts with equal bids will be popped in a random order is better.

File: arbitrum-docs/stylus/concepts/stylus-cache-manager.md (hunk 1 above)
Comment (tsahee) [inline arbitrum-docs/stylus/concepts/stylus-cache-manager.md:31]: Here usage of "activation" is correct, and "initialization" is used to describe the thing that the cache reduces the cost of.

**Review by anegg0 — APPROVED** [review]
Looks great to me, I did minor edits, feel free to let me know if you object to any of them.

**Review by anegg0 — APPROVED** [review]

**Pull request merged**